include/clang/AST/Expr.h

```diff
@@ -955,9 +955,43 @@ class DeclRefExpr final
       private llvm::TrailingObjects<DeclRefExpr, NestedNameSpecifierLoc,
                                     NamedDecl *, ASTTemplateKWAndArgsInfo,
                                     TemplateArgumentLoc> {
+public :
+  /// \brief extra information for DeclRefExprInfo.
+  class GenericFunctionCallInfo {
+  public:
+    struct TypeNameInfo {
+      QualType typeName;
+      TypeSourceInfo *sourceInfo;
+    };
+
+  private :
+    /// \brief If DeclRefExpr references generic function, it must store
```

Comment:
I suggest rewording the comment to be more precise: references to generic functions in code must always be instantiated (applied to type arguments)  We store the type arguments in DeclRefExpr.

Instead of calling the member TypeNameInfos, I suggest TypeArguments.